src/Embedders/Word2Vec.php

```diff
@@ -0,0 +1,790 @@
+
+        foreach ($samples as &$sample) {
+            foreach ($sample as $i2 => $sentence) {
+                $preppedSentence = $this->prepSentence($sentence);
+
+                $embeddings = [];
+                foreach ($preppedSentence as $word) {
+                    $embeddings[] = $this->embedWord($word);
+                }
+            }
+
+            if (isset($embeddings)) {
+                $sample = Matrix::stack($embeddings)->transpose()->mean()->asArray();
+            }
+        }
+    }
+
+    /**
+     * Determine the top N similar words provided an array of positive words and negative words.
+     *
+     * @param string[] $positive
+     * @param string[] $negative
+     * @param int $top
+     * @return string[] $result
+     */
+    public function mostSimilar(array $positive, array $negative = [], $top = 20) : array
+    {
+        $positiveArray = $negativeArray = $allWords = $means = [];
+
+        foreach ($positive as $word) {
+            $positiveArray[$word] = 1.0;
+        }
+
+        foreach ($negative as $word) {
+            $negativeArray[$word] = -1.0;
+        }
+
+        $wordArray = array_merge($positiveArray, $negativeArray);
+
+        foreach ($wordArray as $word => $weight) {
+            $wordEmbedding = $this->wordVec($word);
+
+            if ($wordEmbedding instanceof Vector) {
+                $means[] = $wordEmbedding->multiplyScalar($weight);
+                $allWords[] = $this->vocab[$word]['index'];
+            }
+        }
+
+        if (empty($allWords)) {
+            throw new InvalidArgumentException('Positive words were not found in vocab.');
+        }
+
+        $mean = Matrix::stack($means)->transpose()->mean();
+        $l2 = Matrix::stack($this->vectorsNorm);
+        $dists = $mean->transpose()->matmul($l2->transpose())->asArray()[0];
+
+        arsort($dists, SORT_REGULAR);
+
+        $result = [];
+        foreach ($dists as $index => $weight) {
+            if (!in_array($index, $allWords)) {
+                $result[$this->index2word[$index]] = $weight;
+            }
+        }
+
+        return array_slice($result, 0, $top, true);
+    }
+
+    /**
+     * Update vector weights and return error.
+     *
+     * @param mixed[] $predictWord
+     * @param \Tensor\Vector $l1
+     * @return \Tensor\Vector
+     */
+    protected function error($predictWord, $l1) : Vector
+    {
+        $wordIndices = $this->approximation->wordIndices($predictWord);
+        $l2 = $this->layerMatrix($wordIndices);
+        $fa = $this->propagateHidden($l2, $l1);
+        $gd = $this->approximation->gradientDescent($fa, $predictWord['word'], $this->alpha);
+
+        $this->learnHidden($wordIndices, $gd, $l1);
+
+        return $this->error->addMatrix($gd->matmul($l2))->rowAsVector(0);
+    }
+
+    /**
+     * Scan vocab, prepare vocab, sort vocab, set sampling methods
+     *
+     * @param string[] $sentences
+     */
+    private function preprocess(array $sentences) : void
```

Comment:
The question is do we handle pre-processing in the constructor or in fit?

The "preprocess" method handles all vocabulary preprocessing so we can move it wherever we need to. The reason it's currently in the fit method is because the fit contract requires the Dataset parameter. So we don't have the information required to preprocess until the fit method is called.